### PR TITLE
feat(pubsub): introduce PublisherOptionList (g::c::Options)

### PR DESCRIPTION
**Pull request by dbolduc**
(Continuation of #7273) I think I was supposed to reopen the closed PR then push to it, instead of force-pushing when it was closed.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7308)

### CODE REVIEW

File: google/cloud/pubsub/publisher_options.h [hunk 1]
```diff
@@ -176,73 +205,73 @@ class PublisherOptions {
...
+  Options opts_;
 };
 
 }  // namespace GOOGLE_CLOUD_CPP_PUBSUB_NS
 }  // namespace pubsub
+
+namespace pubsub_internal {
+inline namespace GOOGLE_CLOUD_CPP_PUBSUB_NS {
+
+inline Options&& MakeOptions(pubsub::PublisherOptions&& o) {
```
Comment (devbww) [inline google/cloud/pubsub/publisher_options.h:269]: Is return by rref needed, or even good?

**Review by dbolduc — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7308)* status: 0 of 12 files reviewed, 6 unresolved discussions (waiting on @coryan and @devbww)

___
*[google/cloud/pubsub/options.h, line 44 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7308#-MjV1IEo70uDXDmZdbbB-r1-44:-MjVkeny4u10j2GiS8PK:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/cafc2f52f9d907b600089d75cac3a74c6e7ba28f/google/cloud/pubsub/options.h#L44)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Unused?
</blockquote></details>

Done.
___
*[google/cloud/pubsub/options.h, line 54 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7308#-MjV1IEo70uDXDmZdbbB-r1-54:-MjVzXgl-pZo2-PcgQgn:b-dg692n) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/cafc2f52f9d907b600089d75cac3a74c6e7ba28f/google/cloud/pubsub/options.h#L54)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Nit: Some notes have initial caps, and some don't.
</blockquote></details>

Fixed in this PR. And: https://github.com/googleapis/google-cloud-cpp/pull/7310
___
*[google/cloud/pubsub/publisher.cc, line 24 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7308#-MjV1IEo70uDXDmZdbbC-r1-24:-MjWhEEf8iCWX-BoaPa9:b-ny8x01) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/cafc2f52f9d907b600089d75cac3a74c6e7ba28f/google/cloud/pubsub/publisher.cc#L24)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

I assume that at some point in the future these will be moved somewhere?
</blockquote></details>

I am not sure I follow what you are asking.

I assume that there is currently a `PublisherOptions` param for backwards compatibility, which is now ignored because the `PublisherOptions` have already been used to configure the `PublisherConnection`.

I think at some point in the future `PublisherOptions` will be deprecated, and we will just make a new constructor that only accepts a `std::shared_ptr<PublisherConnection> connection`.

did that answer your question?
___
*[google/cloud/pubsub/publisher_options.h, line 269 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7308#-MjV1IEo70uDXDmZdbbE-r1-269:-MjW-aa-2Qsy7_5u2FoE:bsajjm7) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/cafc2f52f9d907b600089d75cac3a74c6e7ba28f/google/cloud/pubsub/publisher_options.h#L269)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Is return by rref needed, or even good?
</blockquote></details>

It is definitely not needed. Based on the types of the options, I don't think it adds anything. Fixed.
___
*[google/cloud/pubsub/internal/defaults.h, line 26 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7308#-MjV1IEo70uDXDmZdbb8-r1-26:-MjWexA_5p8YeNJIS3Ky:bt2vr1b) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/cafc2f52f9d907b600089d75cac3a74c6e7ba28f/google/cloud/pubsub/internal/defaults.h#L26)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

If the only place you envisage defaulting the argument is in the test, then I'd remove the default.
</blockquote></details>

Fixed.
___
*[google/cloud/pubsub/internal/defaults.cc, line 37 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7308#-MjV1IEnQkrCk3y29m_n-r1-37:-MjVp6om2Mmavc7jijca:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/cafc2f52f9d907b600089d75cac3a74c6e7ba28f/google/cloud/pubsub/internal/defaults.cc#L37)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

`#include <limits>`
</blockquote></details>

Done.

**Review by devbww — CHANGES_REQUESTED** [review]
> Reviewed 6 of 12 files at r1, 6 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7308)* status: all files reviewed, 2 unresolved discussions (waiting on @coryan and @dbolduc)

___
*[google/cloud/pubsub/publisher.cc, line 24 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7308#-MjV1IEo70uDXDmZdbbC-r1-24:-MjXtugH6RWd0MI8DY0b:b-uck9qi) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/cafc2f52f9d907b600089d75cac3a74c6e7ba28f/google/cloud/pubsub/publisher.cc#L24)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

I am not sure I follow what you are asking.

I assume that there is currently a `PublisherOptions` param for backwards compatibility, which is now ignored because the `PublisherOptions` have already been used to configure the `PublisherConnection`.

I think at some point in the future `PublisherOptions` will be deprecated, and we will just make a new constructor that only accepts a `std::shared_ptr<PublisherConnection> connection`.

did that answer your question?

</blockquote></details>

If the parameter is unused, can we just make it a `const&` and avoid a copy and a `NOLINT`?
___
*[google/cloud/pubsub/publisher_options_test.cc, line 113 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7308#-MjXvAQk9u9b8c2rEX-A:-MjXvAQlBb5xSTyu-Lnm:b-tgxw1r) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/688e298a7965f177f79aead7582e7c72ff27ce6c/google/cloud/pubsub/publisher_options_test.cc#L113)):*
> ```cpp
>                .enable_message_ordering();
> 
>   auto opts = pubsub_internal::MakeOptions(b);
> ```

It seems like it is still a benefit to move these `PublisherOptions` into the argument if the local variable is otherwise dead.

**Review by dbolduc — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7308)* status: 10 of 13 files reviewed, 2 unresolved discussions (waiting on @coryan and @devbww)

___
*[google/cloud/pubsub/publisher.cc, line 24 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7308#-MjV1IEo70uDXDmZdbbC-r1-24:-Mj_-7l052gP5PT7qnEi:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/cafc2f52f9d907b600089d75cac3a74c6e7ba28f/google/cloud/pubsub/publisher.cc#L24)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

If the parameter is unused, can we just make it a `const&` and avoid a copy and a `NOLINT`?
</blockquote></details>

Done.
___
*[google/cloud/pubsub/publisher_options_test.cc, line 113 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7308#-MjXvAQk9u9b8c2rEX-A:-Mj_-8gPBC9JDLhZoDou:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/688e298a7965f177f79aead7582e7c72ff27ce6c/google/cloud/pubsub/publisher_options_test.cc#L113)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

It seems like it is still a benefit to move these `PublisherOptions` into the argument if the local variable is otherwise dead.
</blockquote></details>

Done.

**Review by devbww — APPROVED** [review]
> Reviewed 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7308)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @coryan)

**Pull request merged**